OTLP traces export errors use a consistent error message prefix

jmacd opened the pull request:
When a user configures both a tracing SDK and metrics SDK the, existing error-handling and logging mechanism do not indicate whether it is a tracing or a metrics failure. For SDKs and vendors that route metrics and traces to different services, it is important to indicate which of the two paths is experiencing problems.

Part of https://github.com/open-telemetry/opentelemetry-go/issues/3513.

MrAlias commented:
Please fix the lint issues and add a changelog entry.

MrAlias reviewed (commented):
This seems like a lot of code to add to ensure the two client error messages have a prefix. Why not just add the prefix to the client response?

dashpole commented on exporters/otlp/otlptrace/otlptracehttp/client_test.go:
```diff
@@ -219,7 +218,8 @@ func TestTimeout(t *testing.T) {
 		assert.NoError(t, exporter.Shutdown(ctx))
 	}()
 	err = exporter.ExportSpans(ctx, otlptracetest.SingleReadOnlySpan())
-	assert.Equalf(t, true, os.IsTimeout(err), "expected timeout error, got: %v", err)
```
Can we also keep this assertion that we can unwrap error?

jmacd replied:
It's difficult to preserve the behavior of `os.IsTimeout()`, actually, when you wrap an error. I think we can do is create a struct to wrap errors and prefix in the `Error()` method, we have to avoid `fmt.Errorf()`.

dashpole reviewed (approved):
Overall, I think adding information to the the error message is helpful, and worth the extra branch.

jmacd commented:
> This seems like a lot of code to add to ensure the two client error messages have a prefix. 

This point is well taken. This PR and #3515 can be done better and I will give it a try. My biggest concern among these three is actually #3514, which conceals the gRPC or HTTP error message when the final error was retryable. I'll close this and #3515 for now and reopen when it's in better shape.

> Why not just add the prefix to the client response?

There are cases where the retry mechanism returns a local error that would not be prefixed, that's what led me to this in the first place. When you get the full gRPC error message you're likely to have enough context to determine whether it's a tracing SDK problem or a metrics SDK problem.

jmacd commented on exporters/otlp/otlptrace/otlptracegrpc/client_test.go:
```diff
@@ -239,7 +237,8 @@ func TestExportSpansTimeoutHonored(t *testing.T) {
 	// Release the export so everything is cleaned up on shutdown.
 	close(exportBlock)
 
-	require.Equal(t, codes.DeadlineExceeded, status.Convert(err).Code())
```
I will rework this PR. Thanks for asking :grin:

jmacd commented:
I changed this PR to use a wrapped error and restored the old tests w/ a call to `errors.Unwrap()`, since if you want to use `os.IsTimeout()` or the `grpc/status` package you need to have the exact error. This makes it a slight change for callers, but I've updated all the tests and the changelog to indicate that `errors.Unwrap` is needed.

jmacd commented:
If the approach taken here is acceptable, I would do the same (using the new internal `wrappedExportError` type from this PR) for metrics. Thanks.

dashpole commented on exporters/otlp/internal/wrappederror.go:
```diff
@@ -0,0 +1,61 @@
+
+// WrapTracesError wraps an error from the OTLP exporter for traces.
+func WrapTracesError(err error) error {
+	return wrappedExportError{
```
Is this just the same as `return fmt.Errorf("traces export: %w")`?

jmacd replied:
It's pretty close. The call to `errors.Unwrap()` does the same thing, but this would also allow code at the exporter/otlp level to inspect errors.
